### PR TITLE
Tdl 24072 fix typecasting bug max api limit

### PR DESCRIPTION
**Pull request by Vi6hal**
# Description of change
- Fixed ```int() argument must be a string, a bytes-like object or a number, not 'NoneType'``` issue with apI_max_limit typecasting

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_marketo/client.py [hunk 1]
```diff
@@ -89,7 +89,13 @@ def __init__(self, endpoint, client_id, client_secret,
         self.domain = extract_domain(endpoint)
         self.client_id = client_id
         self.client_secret = client_secret
-        self.max_daily_calls = int(max_daily_calls)
+        try:
+            self.max_daily_calls = int(max_daily_calls or MAX_DAILY_CALLS)
+        except (ValueError, TypeError):
+            # Raises TypeError for None / Null Value
+            # Raises ValueError for "" / bool value
```
Comment (dsprayberry) [inline tap_marketo/client.py:96]: can we make this a logger error like:

```LOGGER.warning("Invalid max_daily_calls was passed: \'{}\', using default value of 40,000.".format("max_daily_calls"))```

That way it's a little clearer what might have happened to folks reading the logs

File: tests/test_client_max_api_limit.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import unittest
+from datetime import datetime,timedelta
+from tap_marketo.client import Client, MAX_DAILY_CALLS
+
+class TestDateWindowConfig(unittest.TestCase):
+    
+    def test_datewindow_disabled_no_val(self):
```
Comment (dsprayberry) [inline tests/test_client_max_api_limit.py:7]: These need to be renamed since it's not using datewindow

**Review by dsprayberry — CHANGES_REQUESTED** [review]

File: tap_marketo/client.py (hunk 1 above)
Comment (RushiT0122) [reply tap_marketo/client.py:96]: `int(max_daily_calls or MAX_DAILY_CALLS)` expression will fallback to default value for `None`, `True`/`False` and `""`.

IMO any other exception should be raised instead of falling back to default value.

```
>>> int("" or 100)
100
>>> int(True or 100)
1
>>> int(False or 100)
100
>>> int('1000' or 100)
1000
>>> int('True' or 100)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'True'
```

File: tests/test_client_max_api_limit.py (hunk 2 above)
Comment (Vi6hal) [reply tests/test_client_max_api_limit.py]: updated this

File: tap_marketo/client.py (hunk 1 above)
Comment (Vi6hal) [reply tap_marketo/client.py:96]: I don't like the idea of failing the code for a non-mandatory value but suppressing it will not handle all scenarios such as, this will only result in the user observing unexpected behaviour.
- Number with "," eg: 1,000,000 (internationalisation issues)
- Negative Number
- Decimal input
- Zero as a string i.e "0" (which will result in the limit being set to Zero).

Hence i find it best to raise an error incase of any such values instead of using a fallback mechanism.

Comment (Vi6hal) [reply tap_marketo/client.py:96]: the UI for this tap also suggested a value with a comma for the api_limit i.e `50,000` which is bound to fail.

image

Comment (Vi6hal) [reply tap_marketo/client.py:96]: i have added a few more test scenarios and handled the negative / zero limit condition

Comment (dsprayberry) [reply tap_marketo/client.py]: We have precedence in other taps of logging if we're going to fall back to a default value and allowing the tap to try to proceed with the default, but not really for an API quota scenario, so I'm good either way

**Review by dsprayberry — APPROVED** [review]

File: tests/test_client_max_api_limit.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+from tap_marketo.client import Client, MAX_DAILY_CALLS
+
+class TestmaxdailycallsConfig(unittest.TestCase):
+    
+    def test_maxdailycalls_default_no_val(self):
+        """
+            Verify that max_daily_calls is default if no value is passed
+        """
+        # Initialize Client object
+        client = Client(**{'endpoint': "123-ABC-789",'client_id':'ABC-123','client_secret':'123-QRT'})
```
Comment (RushiT0122) [inline tests/test_client_max_api_limit.py:11]: Be consistent with the `'` and `"`

File: tests/test_client_max_api_limit.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        
+        self.assertEqual(client.max_daily_calls, MAX_DAILY_CALLS)
+
+    def test_maxdailycalls_default_bool_val(self):
+        """
+            Verify that max_daily_calls is default if bool value is passed
+        """
+        # Initialize Client object
+        client = Client(**{'endpoint': "123-ABC-789",'client_id':'ABC-123','client_secret':'123-QRT','max_daily_calls':False})
+        self.assertEqual(client.max_daily_calls ,MAX_DAILY_CALLS)
```
Comment (RushiT0122) [inline tests/test_client_max_api_limit.py:31]: `True` bool value should also be tested.

**Comment by Vi6hal** [comment]
```python
try:
     self.max_daily_calls = int(max_daily_calls or MAX_DAILY_CALLS)
     if self.max_daily_calls <= 1:
        raise ValueError("Limit Cannot be Negative or Zero")
except (ValueError, TypeError) as err:
    singer.log_critical(f"Invalid Value passed for max_daily_calls: {max_daily_calls}")
     raise err
```
This is the only issue i see with the current implementation, is this 👇🏼 

If the value is passed as a string "0" it throws an exception since the value is negative/zero, but if the same value is passed as 0 it sets it to default.

eg:
```python
>>> int("0" or 40000)
0
>>> int(0 or 40000)
40000
```
